ssh/tailssh/tailssh.go

```diff
@@ -312,75 +331,107 @@ func (srv *server) handleSessionTermination(ctx context.Context, s ssh.Session,
 func (srv *server) isActiveSession(sctx ssh.Context) bool {
 	srv.mu.Lock()
 	defer srv.mu.Unlock()
-	return srv.activeSessions[sctx.SessionID()]
+	_, ok := srv.activeSessionByH[sctx.SessionID()]
+	return ok
 }
 
-// startSession registers s as an active session.
-func (srv *server) startSession(s ssh.Session) {
+// startSession registers ss as an active session.
+//
+// It also sets ss.sharedID.
+func (srv *server) startSession(ss *sshSession) {
 	srv.mu.Lock()
 	defer srv.mu.Unlock()
-	if srv.activeSessions == nil {
-		srv.activeSessions = make(map[string]bool)
+	if srv.activeSessionByH == nil {
+		srv.activeSessionByH = make(map[string]*sshSession)
+	}
+	if srv.activeSessionBySharedID == nil {
+		srv.activeSessionBySharedID = make(map[int64]*sshSession)
+	}
+	if ss.idH == "" {
+		panic("empty id")
+	}
+	if _, dup := srv.activeSessionByH[ss.idH]; dup {
+		panic("dup H id")
+	}
+	srv.activeSessionByH[ss.idH] = ss
+	for {
+		sharedID := time.Now().UnixNano()
+		if _, ok := srv.activeSessionBySharedID[sharedID]; !ok {
+			ss.sharedID = sharedID
+			srv.activeSessionBySharedID[sharedID] = ss
+			return
+		}
 	}
-	srv.activeSessions[s.Context().(ssh.Context).SessionID()] = true
 }
 
 // endSession unregisters s from the list of active sessions.
-func (srv *server) endSession(s ssh.Session) {
+func (srv *server) endSession(ss *sshSession) {
 	srv.mu.Lock()
 	defer srv.mu.Unlock()
-	delete(srv.activeSessions, s.Context().(ssh.Context).SessionID())
+	delete(srv.activeSessionByH, ss.idH)
+	delete(srv.activeSessionBySharedID, ss.sharedID)
 }
 
-// handleAcceptedSSH handles s once it's been accepted and determined
-// that it should run as local system user lu.
+var errSessionDone = errors.New("session is done")
+
+// run is the entrypoint for a newly accepted SSH session.
 //
 // When ctx is done, the session is forcefully terminated. If its Err
 // is an SSHTerminationError, its SSHTerminationMessage is sent to the
 // user.
-func (srv *server) handleAcceptedSSH(ctx context.Context, s ssh.Session, ci *sshConnInfo, lu *user.User) {
-	srv.startSession(s)
-	defer srv.endSession(s)
+func (ss *sshSession) run() {
```

Comment:
keep handleAcceptedSSH so that it can munge with server level data and have it call ss.run.

Reply:
It's awkward with the special context type. I added a constructor instead which helps the tests and breaks up the caller. How's that look?